File: sdk/src/main/java/com/meniga/sdk/converters/MenigaFeedConverter.java
package com.meniga.sdk.converters;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import com.google.gson.reflect.TypeToken;
import com.meniga.sdk.helpers.FeedItemFactory;
import com.meniga.sdk.models.feed.MenigaFeed;

import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.annotation.Annotation;
import java.lang.reflect.Type;

import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.Converter;
import retrofit2.Retrofit;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class MenigaFeedConverter extends MenigaConverter {
	private FeedItemFactory feedItemFactory;

	public MenigaFeedConverter(FeedItemFactory feedItemFactory) {
		this.feedItemFactory = feedItemFactory;
	}

	@Override
	public Converter<ResponseBody, ?> responseBodyConverter(Type type, Annotation[] annotations, Retrofit retrofit) {
		Type typeOfFeed = new TypeToken<MenigaFeed>() {}.getType();

		if (typeOfFeed.equals(type)) {
			return new Converter<ResponseBody, MenigaFeed>() {
				@Override
				public MenigaFeed convert(ResponseBody resBody) throws IOException {
					MenigaFeed feed = new MenigaFeed();

					JsonArray arr = getAsArray(resBody.byteStream());
					for (JsonElement element : arr) {
						feed.add(feedItemFactory.getMenigaFeetItem((JsonObject) element));
					}

					try {
						InputStreamReader isr = new InputStreamReader(resBody.byteStream());
						JsonElement jelement = new JsonParser().parse(isr);
						JsonObject jobject = jelement.getAsJsonObject();
						if (jobject.has("meta")) {
							JsonObject meta = jobject.getAsJsonObject("meta");
							feed.setHasMoreData(meta.get("hasMoreData").getAsBoolean());
						} else {
							feed.setHasMoreData(true);
						}
					} catch (Exception ex) {
						feed.setHasMoreData(true);
					}

					return feed;
				}
			};
		}
		return null;
	}

	@Override
	public Converter<?, RequestBody> requestBodyConverter(Type type, Annotation[] parameterAnnotations, Annotation[] methodAnnotations, Retrofit retrofit) {
		return null;
	}
}

File: sdk/src/main/java/com/meniga/sdk/converters/MenigaFeedConverter.kt
/*
 * Copyright 2017-2018 Meniga Iceland Inc.
 */
package com.meniga.sdk.converters

import com.google.gson.Gson
import com.google.gson.JsonObject
import com.meniga.sdk.helpers.FeedItemFactory
import com.meniga.sdk.helpers.get
import com.meniga.sdk.helpers.type
import com.meniga.sdk.models.accounts.MenigaAccount
import com.meniga.sdk.models.feed.MenigaFeed
import com.meniga.sdk.models.feed.setActualEndDate
import com.meniga.sdk.models.feed.setHasMoreData
import com.meniga.sdk.models.feed.setTotalCount
import com.meniga.sdk.models.merchants.MenigaMerchant
import okhttp3.ResponseBody
import org.joda.time.DateTime
import retrofit2.Converter
import retrofit2.Retrofit
import java.lang.reflect.Type

class MenigaFeedConverter(
        private val feedItemFactory: FeedItemFactory,
        private val gson: Gson
) : MenigaConverter() {

    private val transactionEventSorter = TransactionEventSorter()

    override fun responseBodyConverter(type: Type?, annotations: Array<Annotation>?, retrofit: Retrofit?): Converter<ResponseBody, *>? {
        return when (type) {
            type<MenigaFeed>() -> Converter<ResponseBody, MenigaFeed> { resBody ->
                MenigaFeed().apply {
                    val (data, meta, included) = getAsArrayApiResponse(resBody.byteStream())
                    val (accounts, merchants) = gson.getAccountsAndMerchants(included)

                    data.forEach {
                        add(feedItemFactory.getMenigaFeedItem(it as JsonObject, accounts, merchants))
                    }

                    transactionEventSorter.moveTransactionEventsToTransaction(this)

                    meta?.let {
                        val actualEndDate = it.get("actualEndDate")
                        if (!actualEndDate.isJsonNull) {
                            setActualEndDate(DateTime.parse(actualEndDate.asString))
                        }
                        setHasMoreData(it.get("hasMoreData").asBoolean)
                        setTotalCount(it.get("totalCount").asInt)
                    }
                }
            }
            else -> null
        }
    }

    private fun Gson.getAccountsAndMerchants(included: JsonObject?)
            : Pair<List<MenigaAccount>, List<MenigaMerchant>> = Pair(
            included?.let { get<List<MenigaAccount?>>(it, "accounts") }
                    .orEmpty()
                    .filterNotNull(),
            included?.let { get<List<MenigaMerchant?>>(it, "merchants") }
                    .orEmpty()
                    .filterNotNull()
    )
}
